### PR TITLE
Plug YJIT memory leaks

### PR DESCRIPTION
**Pull request by XrXr**
Please check each commit.
Since deallocation is tied to global invalidation, there is a sizeable memory saving for
railsbench as it uses Zeitwerk, which uses TracePoint during eager loading:

|        | `yjit_alloc_size` |
|:------:|:---------------:|
| before |     15287405    |
|  after |     11746945    |

PS: I would prefer to merge this as separate commits (i.e. not squashed together). 🙇

Fixes: Shopify/ruby#463

Memory usage is stable using the test from the original report:
```
triggering code GC
code_gc_count: 1
live_page_count: 2
yjit_alloc_size: 334,313 bytes

triggering code GC
code_gc_count: 2
live_page_count: 2
yjit_alloc_size: 334,409 bytes

...

triggering code GC
code_gc_count: 344
live_page_count: 2
yjit_alloc_size: 334,409 bytes
```

### CODE REVIEW

File: yjit/src/core.rs [hunk 1]
```diff
@@ -2087,12 +2090,7 @@ pub fn defer_compilation(
     asm.mark_branch_end(&branch_rc);
 }
 
-// Remove all references to a block then free it.
-pub fn free_block(blockref: &BlockRef) {
-    use crate::invariants::*;
-
-    block_assumptions_free(blockref);
-
+fn yank_from_control_flow_graph(blockref: &BlockRef) {
```
Comment (maximecb) [inline yjit/src/core.rs:2093]:
```suggestion
fn remove_block_from_graph(blockref: &BlockRef) {
```

File: yjit/src/core.rs [hunk 2]
```diff
@@ -492,6 +492,9 @@ pub struct IseqPayload {
 
     // Indexes of code pages used by this this ISEQ
     pub pages: HashSet<usize>,
+
+    // Blocks that are invalidated but are not yet deallocated
+    pub zombie_blocks: Vec<BlockRef>,
```
Comment (maximecb) [inline yjit/src/core.rs:497]:
```suggestion
    // Blocks that are invalidated but are not yet deallocated
    // The code GC will free these later
    pub dead_blocks: Vec<BlockRef>,
```

File: yjit/src/core.rs [hunk 3]
```diff
@@ -2293,14 +2304,39 @@ pub fn invalidate_block_version(blockref: &BlockRef) {
...
+//   p foo(1)
+pub fn delayed_deallocation(blockref: &BlockRef) {
+    block_assumptions_free(blockref);
+
+    // We do this another time when we deem that it's safe
+    // to deallocate in case there is another Ractor waiting to acquire the
+    // VM lock inside branch_stub_hit().
+    yank_from_control_flow_graph(blockref);
+
+    let payload = get_or_create_iseq_payload(blockref.borrow().blockid.iseq);
```
Comment (maximecb) [inline yjit/src/core.rs:2336]: Minor point, but shouldn't the iseq payload always exist here, or is there a special case where it doesn't? :O

**Review by maximecb — COMMENTED** [review]
I made some minor comments but it looks good. Thanks a lot for fixing this leak Alan, and it's great that memory usage is reduced too!

**Review by maximecb — APPROVED** [review]

**Review by k0kubun — APPROVED** [review]
Nice :clap:

**Pull request merged**